Add ADC channel configuration on ADC-only pins

tiagojbalmeida opened the pull request:
This pull request defines the possibility to configure ADC channel on ADC-only pins. Currently ADC channel is configured on multiplexed pins whose analog function is selected through the ENAIO register in the SCU. However, when we want for instance to use ADC-only pins for thermistor and their correspondent multiplexed pins for something else, we will have issues through this ENAIO register. 

Now, my idea is to change this register when we only configure a multiplexed pin on config file and remain unchanged when we only configure an ADC-only pin.

Example:

Considering we want to connect thermistor pin on ADC0_0:
 
- If we want to configure thermistor on ADC-only pin:
hotend.thermistor_pin = ADC0_0     # Pin for the thermistor to read\n\

- If we want to configure thermistor on multiplexed pin with analog function:
hotend.thermistor_pin = p4_3     # Pin for the thermistor to read\n\

Commit:
Add ADC channel configuration on ADC-only pins

wolfmanjm commented:
This makes no sense. now an adc pin will not get properly configured. if it is and adc0. so rejected.

tiagojbalmeida commented:
We tried to run smoothie in a custom board, where we connected thermistor on channel 0 of ADC-only pin, and P4_3 was connected to a led. Even if we configure ADC0_0, we couldn't get thermistor temps because this ENAIO register enabled p4_3 analog function and led slightly turned on, when it shouldn't. This was a simple solution i could find to differentiate both pins and fixed this issue. The p4_3 can be still configured as ADC0_0 if we wish and ADC-only pin ADC0_0 does not need this register to function properly as datasheet mentions it.

arthurwolf commented:
What would be a correct way of implementing this ? Ultimately we do want to
be able to use the ADC-only pins as some of the board designs use these.

On Thu, Mar 15, 2018 at 1:59 PM, tiagojbalmeida <notifications@github.com>
wrote:

> We tried to run smoothie in a custom board, where we connected thermistor
> on channel 0 of ADC-only pin, and P4_3 was connected to a led. Even if we
> configure ADC0_0, we couldn't get thermistor temps because this ENAIO
> register enabled p4_3 analog function and led slightly turned on, when it
> shouldn't. This was a simple solution i could find to differentiate both
> pins and fixed this issue. The p4_3 can be still configured as ADC0_0 if we
> wish and ADC-only pin ADC0_0 does not need this register to function
> properly as datasheet mentions it.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/smoothie-v2/pull/2#issuecomment-373367196>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFVxLv-Wh0nk7t6pIntoDt6HqZ1waks5temWfgaJpZM4SsAPG>
> .
>



-- 
Courage et bonne humeur.

tiagojbalmeida commented:
I believe this is the correct way. According to the actual firmware, for instance if you want to configure ADC0_0 as a ADC-only pin, you cannot enable a digital function on P4_3 pin, because the function controlling ENAIO register enables analog function ADC0_0 for P4_3 pin. This works perfectly if we do use either ADC0_0 pin or P4_3 pin as ADC, however this would limit the amount of pins and also your board design, since we cannot connect something else (e.g. led) on that P4_3 pin. My PR provides the use of both pins independently as ADC, ADC0_0 would enable the ADC-only pin and P4_3 would enable the multiplexed pin with analog function. Maybe i'm missing something, but my tests did work as expected.

wolfmanjm commented:
The thing is Chip_SCU_PinMuxSet() is what sets the mux, the Chip_SCU_ADC_Channel_Config() enables the channel. So I think your code is wrong.

tiagojbalmeida commented:
I agree, but somehow the ADC-only pins worked without enabling the channel, i think the channel is enabled by default. 

LPC43xx User Manual provides this information on section 17.4.6 "ADC0 function select register":

" For pins with digital and analog functions, this register selects the input channel of the
ADC0 over any of the possible digital functions. This option is not available for channel
ADC0_7.

In addition, each analog function is pinned out on a dedicated analog pin which is not
affected by this register. "

wolfmanjm commented:
Ok I need to read the register specs again to see if I can understand this. AFAIK that call is to enable an ADC channel and has nothing to do with MUX pins. It does not make sense to not enable the channel, and to have to do a MUX and a channel select for dual function pins. If this register overrides the mux then why does it need the mux?

tiagojbalmeida commented:
The datasheet clearly states that the ENAIO register controlled by this function i moved up selects the input ADC channel OVER a digital function (enabled by default) of a multiplexed digital pin. Otherwise, if we configure a dedicated analog pin as ADC pin, this is not affected by this register because it contains ONLY the analog function. That is why i moved this function into the mux pins instruction which i believed it's its place since it will configure and select the analog function of this pin.

The problem/bug i have seen was if you configure an analog ADC pin, this call would enable the analog function of the correspondent mux pin, but if you're using this mux pin as a digital pin, it will not work properly as a digital one.

Anyways, I think the function which enables ADC channel is Chip_ADC_EnableChannel() which controls the SEL symbol of the CR register.
